docs/design/2019-04-11-indexmerge.md

```diff
@@ -0,0 +1,320 @@
+In present TiDB, a SQL statement with conditions involving multiple indexed attributes only uses one of the conditions as the index filter to build access condition, while others are regarded as table filters. Firstly, use index scan (at most one index) to get handles (rowid in TiDB). Then use the handles to get rows and check whether the rows satisfy the conditions of table filters. Some relational databases implement a table access path using multiple indexes. In some cases, this way will improve performance.
+
+We take an example to explain it. We define the table schema as :
+
+```
+	CREATE TABLE t1 (a int, b int, c int);
+	CREATE INDEX t1a on t1(a);
+	CREATE INDEX t1b on t1(b);
+	CREATE INDEX t1c on t1(c);
+```
+And use a test SQL statement `SELECT * FROM t1 where a < 2 or b > 50`. Currently, TiDB does a table scan  and puts `a < 2 or b > 50` as a Selection on top of it. If the selectivity of `a < 2 ` and `b > 50` is low, a better approach would be using indexes on columns `a` and `b` to retrieve rows respectively, and applying a union operation on the result sets.
+
+
+## Proposal
+In short, we need to consider access paths using multiple indexes.
+
+### Planner
+We propose to add new `IndexMergeReader / PhysicalIndexMergeReader` and `IndexMergeLookUpReader / PhysicalIndexMergeLookUpReader` operators.
+
+Now we just consider the following two kinds of queries: 
+
+（1）Conditions in CNF, e.g, `select * from t1 where c1 and c2 and c3 and …`
+
+In this form, each CNF item can be covered by a single index respectively. For example, if we have single column indexes for `t1.a`, `ta.b` and `t1.c` respectively, for SQL `select * from t1 where (a < 10 or a > 100) and b < 10 and c > 1000`, we can use all the three indexes to read the table handles. The result plan for it is like:
+
+	```
+  	PhysicalIndexMergeLookUpReader(IndexMergeIntersect)
+		IndexScan(t1a)
+		IndexScan(t1b)
+		IndexScan(t1c)
+		TableScan
+	```
+For the CNF items not covered by any index, we take them as table filters and convert them to selection on top of the scan node. For SQL `select * from t1 where (a < 10 or c >100) and b < 10`, only item `b < 10` can be used as index access condition, so we can only consider single index lookup reader.
+
+We set up a experiment for the CNF form to compare our demo implement with the master branch. The schema and test sql form we define are following:
+
+```
+Table Schema:
+	CREATE TABLE T200M(c1 int, c2 int, c3 int, c4 int, c5 int, c6 int, c7 int, c8 int);
+	CREATE INDEX T200Ma on T200M(a);
+	CREATE INDEX T200Mb on T200M(b);
+
+Test SQL Form:	
+	CNF-1
+		SELECT * FROM T200M WHERE C1 < $1 AND C2 > $2;
+	CNF-2 
+		SELECT * FROM T200M WHERE C1 < $3 AND C2 < $4;
+```
+
+We load two million rows into `T200M` with one to two million sequence for all columns. We alter `$1-$4` to obtain equal selectivity on `C1` and `C2`, while the difference between `CNF-1` and `CNF-2` is `CNF-1` has no final results. The result can be seen in the following graph:
```

Comment:
If selectivity for `C1 < $3` is low, while `C2 < $4` is high, single index scan on C1 would be better?